Mathlib/Algebra/BigOperators/Group/Finset/Basic.lean

```diff
@@ -859,163 +853,156 @@ lemma prod_sdiff_ne_prod_sdiff_iff :
+
+lemma sum_tsub_distrib (s : Finset ι) {f g : ι → M} (hfg : ∀ x ∈ s, g x ≤ f x) :
+    ∑ x ∈ s, (f x - g x) = ∑ x ∈ s, f x - ∑ x ∈ s, g x := Multiset.sum_map_tsub _ hfg
+
+end OrderedSub
+
+section Nat
+
+lemma card_eq_sum_ones (s : Finset ι) : #s = ∑ _ ∈ s, 1 := by simp
+
+theorem sum_const_nat {m : ℕ} {f : ι → ℕ} (h₁ : ∀ x ∈ s, f x = m) : ∑ x ∈ s, f x = #s * m := by
+  rw [← Nat.nsmul_eq_mul, ← sum_const]
+  apply sum_congr rfl h₁
+
+lemma sum_card_fiberwise_eq_card_filter {κ : Type*} [DecidableEq κ] (s : Finset ι) (t : Finset κ)
+    (g : ι → κ) : ∑ j ∈ t, #{i ∈ s | g i = j} = #{i ∈ s | g i ∈ t} := by
+  simpa only [card_eq_sum_ones] using sum_fiberwise_eq_sum_filter _ _ _ _
+
 @[simp]
-theorem card_disjiUnion (s : Finset α) (t : α → Finset β) (h) :
+theorem card_disjiUnion (s : Finset ι) (t : ι → Finset M) (h) :
     #(s.disjiUnion t h) = ∑ a ∈ s, #(t a) :=
   Multiset.card_bind _ _
 
-theorem card_biUnion [DecidableEq β] {t : α → Finset β} (h : s.toSet.PairwiseDisjoint t) :
-    #(s.biUnion t) = ∑ u ∈ s, #(t u) := by simpa using sum_biUnion h (β := ℕ) (f := 1)
+theorem card_biUnion [DecidableEq M] {t : ι → Finset M} (h : s.toSet.PairwiseDisjoint t) :
+    #(s.biUnion t) = ∑ u ∈ s, #(t u) := by simpa using sum_biUnion h (M := ℕ) (f := 1)
 
-theorem card_biUnion_le [DecidableEq β] {s : Finset α} {t : α → Finset β} :
+theorem card_biUnion_le [DecidableEq M] {s : Finset ι} {t : ι → Finset M} :
     #(s.biUnion t) ≤ ∑ a ∈ s, #(t a) :=
-  haveI := Classical.decEq α
+  haveI := Classical.decEq ι
   Finset.induction_on s (by simp) fun a s has ih =>
     calc
       #((insert a s).biUnion t) ≤ #(t a) + #(s.biUnion t) := by
         rw [biUnion_insert]; exact card_union_le ..
       _ ≤ ∑ a ∈ insert a s, #(t a) := by rw [sum_insert has]; exact Nat.add_le_add_left ih _
 
-theorem card_eq_sum_card_fiberwise [DecidableEq β] {f : α → β} {s : Finset α} {t : Finset β}
+theorem card_eq_sum_card_fiberwise [DecidableEq M] {f : ι → M} {s : Finset ι} {t : Finset M}
     (H : s.toSet.MapsTo f t) : #s = ∑ b ∈ t, #{a ∈ s | f a = b} := by
   simp only [card_eq_sum_ones, sum_fiberwise_of_maps_to H]
 
-theorem card_eq_sum_card_image [DecidableEq β] (f : α → β) (s : Finset α) :
+theorem card_eq_sum_card_image [DecidableEq M] (f : ι → M) (s : Finset ι) :
     #s = ∑ b ∈ s.image f, #{a ∈ s | f a = b} :=
   card_eq_sum_card_fiberwise fun _ => mem_image_of_mem _
 
-theorem mem_sum {f : α → Multiset β} (s : Finset α) (b : β) :
+theorem mem_sum {f : ι → Multiset M} (s : Finset ι) (b : M) :
     (b ∈ ∑ x ∈ s, f x) ↔ ∃ a ∈ s, b ∈ f a := by
   induction s using Finset.cons_induction with
   | empty => simp
   | cons a t hi ih => simp [sum_cons, ih, or_and_right, exists_or]
 
-@[to_additive]
-theorem prod_unique_nonempty {α β : Type*} [CommMonoid β] [Unique α] (s : Finset α) (f : α → β)
-    (h : s.Nonempty) : ∏ x ∈ s, f x = f default := by
-  rw [h.eq_singleton_default, Finset.prod_singleton]
-
-section Image_Overlap
-
```

Comment:
and here

Reply:
These have moved earlier because they are about commutative monoids